File: backend/services/text_recipes.py
"""Reversible text recipe helpers for QRed seal generation.

The b45 transform is provided by the external ``b45`` package so QRed uses the
same reference implementation as https://github.com/greyphilosophy/b45.
"""

from __future__ import annotations

import base64
from dataclasses import asdict, dataclass

import brotli
from b45 import decode as _b45_decode
from b45 import encode as _b45_encode

B45_RECIPE_ID = "b45"
B45_LONG_RECIPE_ID = "base45ish"
BROTLI_RECIPE_ID = "brotli"


@dataclass(frozen=True)
class RecipeDiagnostic:
    line: int
    reason: str
    original: str = ""
    restored: str = ""
    recommendation: str = ""

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass(frozen=True)
class RecipeValidationResult:
    recipe_id: str
    reversible: bool
    compact: str
    restored: str
    diagnostics: list[RecipeDiagnostic]

    def to_dict(self) -> dict:
        return {
            "recipe_id": self.recipe_id,
            "reversible": self.reversible,
            "compact": self.compact,
            "restored": self.restored,
            "diagnostics": [diag.to_dict() for diag in self.diagnostics],
        }


def encode_b45ish(text: str) -> str:
    return _b45_encode(text)


def decode_b45ish(compact: str) -> str:
    return _b45_decode(compact)


def encode_brotli(text: str) -> str:
    compressed = brotli.compress(text.encode("utf-8"), quality=11)
    return base64.urlsafe_b64encode(compressed).decode("ascii").rstrip("=")


def decode_brotli(compact: str) -> str:
    padded = compact + ("=" * ((4 - (len(compact) % 4)) % 4))
    compressed = base64.urlsafe_b64decode(padded.encode("ascii"))
    return brotli.decompress(compressed).decode("utf-8")


def validate_brotli(original: str) -> RecipeValidationResult:
    try:
        compact = encode_brotli(original)
        restored = decode_brotli(compact)
    except Exception as exc:
        return RecipeValidationResult(
            recipe_id=BROTLI_RECIPE_ID,
            reversible=False,
            compact="",
            restored="",
            diagnostics=[
                RecipeDiagnostic(
                    line=1,
                    reason=f"brotli round-trip failed: {exc}",
                    original=original,
                    recommendation="Use b45 or plaintext instead.",
                )
            ],
        )

    reversible = restored == original
    diagnostics: list[RecipeDiagnostic] = []
    if not reversible:
        diagnostics = [
            RecipeDiagnostic(
                line=1,
                reason="brotli round-trip did not restore the original text.",
                original=original,
                restored=restored,
                recommendation="Use b45 or plaintext instead.",
            )
        ]
    return RecipeValidationResult(
        recipe_id=BROTLI_RECIPE_ID,
        reversible=reversible,
        compact=compact,
        restored=restored,
        diagnostics=diagnostics,
    )


def validate_b45(original: str) -> RecipeValidationResult:
    compact = encode_b45ish(original)
    try:
        restored = decode_b45ish(compact)
    except Exception as exc:
        return RecipeValidationResult(
            recipe_id=B45_RECIPE_ID,
            reversible=False,
            compact=compact,
            restored="",
            diagnostics=[
                RecipeDiagnostic(
                    line=1,
                    reason=f"b45 decoding failed: {exc}",
                    original=original,
                    recommendation="Use plaintext or another reversible recipe instead.",
                )
            ],
        )

    reversible = restored == original
    diagnostics: list[RecipeDiagnostic] = []
    if not reversible:
        diagnostics = [
            RecipeDiagnostic(
                line=1,
                reason="b45 round-trip did not restore the original text.",
                original=original,
                restored=restored,
                recommendation="Use plaintext or another reversible recipe instead.",
            )
        ]
    return RecipeValidationResult(
        recipe_id=B45_RECIPE_ID,
        reversible=reversible,
        compact=compact,
        restored=restored,
        diagnostics=diagnostics,
    )


# Compatibility aliases kept so older call sites keep working while the recipe
# itself lives in the external b45 package.
def encode_simple_english(text: str) -> tuple[str, list[RecipeDiagnostic]]:
    compact = encode_b45ish(text)
    return compact, []


def decode_simple_english(compact: str) -> str:
    return decode_b45ish(compact)


def validate_simple_english(original: str) -> RecipeValidationResult:
    return validate_b45(original)

File: frontend/src/textRecipes.js
const B45_RECIPE_ID = "b45";

// Browser implementation sourced from the b45 reference project demo:
// https://github.com/greyphilosophy/b45/blob/main/docs/demo.html
const QR_ALPHANUMERIC = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:";
const PASS_THROUGH = new Set("0123456789 $.-: ".split(""));
const SHIFT_DECODE_BY_KEY = {
  "1": "!", "2": "@", "3": "#", "6": "^",
  "7": "&", "9": "(", "0": ")", "/": "?", ":": ";",
};
const SHIFT_ENCODE_BY_CHAR = Object.fromEntries(
  Object.entries(SHIFT_DECODE_BY_KEY).map(([key, value]) => [value, key]),
);
const SPECIAL_SINGLE_DECODE_BY_CHAR = { "*": "\"" };
const HEX = new Set("0123456789ABCDEF".split(""));
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder("utf-8", { fatal: true });

function codePointAt(text, index) {
  return String.fromCodePoint(text.codePointAt(index));
}

function encodeUtf8Escape(char, parts) {
  for (const byte of textEncoder.encode(char)) {
    parts.push(`%${byte.toString(16).toUpperCase().padStart(2, "0")}`);
  }
}

function appendPeriodRun(text, index, parts) {
  const start = index;
  while (index < text.length && text[index] === ".") index += 1;
  if (index - start === 1) {
    parts.push(".");
  } else {
    for (let cursor = start; cursor < index; cursor += 1) encodeUtf8Escape(".", parts);
  }
  return index;
}

function encodeApostropheSlashRun(text, index, parts) {
  const start = index;
  while (index < text.length && (text[index] === "'" || text[index] === "/")) index += 1;
  if (index - start === 1) {
    parts.push(text[start] === "'" ? "/" : "//");
  } else {
    for (let cursor = start; cursor < index; cursor += 1) encodeUtf8Escape(text[cursor], parts);
  }
  return index;
}

function encodeChar(char, parts) {
  if (char >= "a" && char <= "z") parts.push(char.toUpperCase());
  else if (char >= "A" && char <= "Z") parts.push(`+${char}`);
  else if (char === "+") parts.push("++");
  else if (char === "%") parts.push("%%");
  else if (char === "\"") parts.push("*");
  else if (char === "*") encodeUtf8Escape(char, parts);
  else if (char === ":") parts.push("::");
  else if (PASS_THROUGH.has(char)) parts.push(char);
  else if (SHIFT_ENCODE_BY_CHAR[char]) parts.push(`+${SHIFT_ENCODE_BY_CHAR[char]}`);
  else encodeUtf8Escape(char, parts);
}

export function encodeB45ish(text) {
  const parts = [];
  let index = 0;
  while (index < text.length) {
    const char = codePointAt(text, index);
    if (char === ",") {
      if (index + 1 < text.length && text[index + 1] === ":") encodeUtf8Escape(char, parts);
      else parts.push(":");
      index += 1;
      continue;
    }
    if (char === ".") {
      index = appendPeriodRun(text, index, parts);
      continue;
    }
    if (char === "'" || char === "/") {
      index = encodeApostropheSlashRun(text, index, parts);
      continue;
    }
    encodeChar(char, parts);
    index += char.length;
  }
  return parts.join("");
}

function decodePlusEscape(encoded, index) {
  if (index + 1 >= encoded.length) throw new Error("Dangling '+' escape at end of input");
  const nextChar = encoded[index + 1];
  if (nextChar === "+") return ["+", index + 2];
  if (nextChar >= "A" && nextChar <= "Z") return [nextChar, index + 2];
  if (SHIFT_DECODE_BY_KEY[nextChar]) return [SHIFT_DECODE_BY_KEY[nextChar], index + 2];
  throw new Error(`Invalid '+' escape at position ${index}: '+${nextChar}'`);
}

function decodeColonEscape(encoded, index) {
  if (index + 1 < encoded.length && encoded[index + 1] === ":") return [":", index + 2];
  return [",", index + 1];
}

function decodeSlashEscape(encoded, index) {
  if (index + 1 < encoded.length && encoded[index + 1] === "/") return ["/", index + 2];
  return ["'", index + 1];
}

function isHexByte(value) {
  return value.length === 2 && [...value].every((char) => HEX.has(char));
}

function decodePercentRun(encoded, index) {
  if (index + 1 < encoded.length && encoded[index + 1] === "%") return ["%", index + 2];
  const bytes = [];
  const start = index;
  while (index < encoded.length && encoded[index] === "%") {
    if (index + 2 >= encoded.length) throw new Error(`Incomplete percent escape at position ${index}`);
    const hexDigits = encoded.slice(index + 1, index + 3);
    if (!isHexByte(hexDigits)) {
      if (index === start) throw new Error(`Invalid percent escape at position ${index}: '%${hexDigits}'`);
      break;
    }
    bytes.push(Number.parseInt(hexDigits, 16));
    index += 3;
  }
  try {
    return [textDecoder.decode(new Uint8Array(bytes)), index];
  } catch {
    throw new Error(`Invalid UTF-8 byte escape run starting at position ${start}`);
  }
}

export function decodeB45ish(encoded) {
  const output = [];
  let index = 0;
  while (index < encoded.length) {
    const char = encoded[index];
    if (!QR_ALPHANUMERIC.includes(char)) throw new Error(`Invalid b45 character at position ${index}: ${JSON.stringify(char)}`);
    if (char === "+") {
      const [decoded, nextIndex] = decodePlusEscape(encoded, index);
      output.push(decoded);
      index = nextIndex;
    } else if (char === ":") {
      const [decoded, nextIndex] = decodeColonEscape(encoded, index);
      output.push(decoded);
      index = nextIndex;
    } else if (SPECIAL_SINGLE_DECODE_BY_CHAR[char]) {
      output.push(SPECIAL_SINGLE_DECODE_BY_CHAR[char]);
      index += 1;
    } else if (char === "/") {
      const [decoded, nextIndex] = decodeSlashEscape(encoded, index);
      output.push(decoded);
      index = nextIndex;
    } else if (char === "%") {
      const [decoded, nextIndex] = decodePercentRun(encoded, index);
      output.push(decoded);
      index = nextIndex;
    } else if (char >= "A" && char <= "Z") {
      output.push(char.toLowerCase());
      index += 1;
    } else if (PASS_THROUGH.has(char)) {
      output.push(char);
      index += 1;
    } else {
      throw new Error(`Invalid b45 character at position ${index}: ${JSON.stringify(char)}`);
    }
  }
  return output.join("");
}

export function validateB45(original) {
  const compact = encodeB45ish(original);
  try {
    const restored = decodeB45ish(compact);
    const reversible = restored === original;
    return {
      recipe_id: B45_RECIPE_ID,
      reversible,
      compact,
      restored,
      diagnostics: reversible ? [] : [{
        line: 1,
        reason: "b45 round-trip did not restore the original text.",
        original,
        restored,
        recommendation: "Use plaintext or legacy compression instead.",
      }],
    };
  } catch (error) {
    return {
      recipe_id: B45_RECIPE_ID,
      reversible: false,
      compact,
      restored: "",
      diagnostics: [{
        line: 1,
        reason: `b45 decoding failed: ${error.message}`,
        original,
        recommendation: "Use plaintext or legacy compression instead.",
      }],
    };
  }
}

// Compatibility aliases for older call sites.
export const encodeSimpleEnglish = encodeB45ish;
export const decodeSimpleEnglish = decodeB45ish;
export const validateSimpleEnglish = validateB45;
